File: scripts/router.ts
/*
 *   Copyright (C) 2020  Joris MASSON
 *
 *   This program is free software: you can redistribute it and/or modify
 *   it under the terms of the GNU Affero General Public License as published by
 *   the Free Software Foundation, either version 3 of the License, or
 *   (at your option) any later version.
 *
 *   This program is distributed in the hope that it will be useful,
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *   GNU Affero General Public License for more details.
 *
 *   You should have received a copy of the GNU Affero General Public License
 *   along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

import Navigo from "navigo";

const SIDEBAR_MENU_APP_LINKS = "#sidebar-menu-app-links";

export const router = new Navigo();

export function initSidebarLinks(doc: Document): Error | null {
    const sidebar_menu = doc.querySelector(SIDEBAR_MENU_APP_LINKS);
    if (!sidebar_menu) {
        return new Error(
            `Could not get the sidebar app menu at id ${SIDEBAR_MENU_APP_LINKS}`
        );
    }

    const sidebar_links = sidebar_menu.querySelectorAll("[data-router-link]");
    for (const link of sidebar_links) {
        if (!(link instanceof HTMLAnchorElement)) {
            return new Error(
                `[data-router-link] should only be applied to <a> tags, got ${link.tagName}`
            );
        }

        const location = link.getAttribute("href");
        if (location === null) {
            return new Error(
                `Could not get the href attribute of a data-router-link <a> tag`
            );
        }

        //TODO: mark the active link with css
        link.addEventListener("click", (event: Event) => {
            event.preventDefault();
            router.navigate(location);
        });
    }

    return null;
}

File: server/adapter/server/app/router.go
/*
 *   Copyright (C) 2020  Joris MASSON
 *
 *   This program is free software: you can redistribute it and/or modify
 *   it under the terms of the GNU Affero General Public License as published by
 *   the Free Software Foundation, either version 3 of the License, or
 *   (at your option) any later version.
 *
 *   This program is distributed in the hope that it will be useful,
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *   GNU Affero General Public License for more details.
 *
 *   You should have received a copy of the GNU Affero General Public License
 *   along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

/*
 Package app groups together app routes. Displaying the frontend app
 to the user with its assets.
*/
package app

import (
	"github.com/gorilla/mux"
	"github.com/hyzual/mike-sierra-sierra/server/adapter"
	"github.com/hyzual/mike-sierra-sierra/server/adapter/server"
	"github.com/hyzual/mike-sierra-sierra/server/adapter/server/user"
	"github.com/swithek/sessionup"
)

// Register registers the routes on the given gorilla/mux router.
func Register(
	router *mux.Router,
	templateExecutor adapter.TemplateExecutor,
	assetsResolver adapter.AssetsResolver,
	userStore user.Store,
	sessionManager *sessionup.Manager,
) {
	appHandler := sessionManager.Auth(
		server.WrapErrors(&appHandler{templateExecutor, assetsResolver, userStore}),
	)
	router.PathPrefix("/app").Handler(appHandler)
}
